### ISSUE TITLE
build: `nogo` can segfault under certain circumstances

### ISSUE THREAD

**Issue by rickystewart** [cockroachdb/cockroach#98260]
[example log](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/8979034?buildTab=overview&showRootCauses=true&expandBuildProblemsSection=true&expandBuildTestsSection=false&expandBuildChangesSection=true)

```
22:55:17       Use --sandbox_debug to see verbose messages from the sandbox
22:55:17       compilepkg: panic: runtime error: invalid memory address or nil pointer dereference
22:55:17       [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5fe13c]
22:55:17       
22:55:17       goroutine 273 [running]:
22:55:17       go/types.methodSet.add(0xc0c880?, {0x0, 0x2, 0xfdc140?}, {0x0, 0x0, 0xfdc140?}, 0x28?, 0xa5?)
22:55:17         GOROOT/src/go/types/methodset.go:214 +0xdc
22:55:17       go/types.NewMethodSet({0xc0c880?, 0xc000ee17a0?})
22:55:17         GOROOT/src/go/types/methodset.go:150 +0x869
22:55:17       golang.org/x/tools/go/types/typeutil.(*MethodSetCache).lookupNamed(0xc0002163b0, 0xc000ee17a0)
22:55:17         golang.org/x/tools/go/types/typeutil/external/org_golang_x_tools/go/types/typeutil/methodsetcache.go:66 +0x99
22:55:17       golang.org/x/tools/go/types/typeutil.(*MethodSetCache).MethodSet(0xc0002163b0, {0xc0c880?, 0xc000ee17a0?})
22:55:17         golang.org/x/tools/go/types/typeutil/external/org_golang_x_tools/go/types/typeutil/methodsetcache.go:37 +0xe5
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c880?, 0xc000ee17a0?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:173 +0xf0
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c970?, 0xc000ef1530?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:242 +0x775
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c8d0?, 0xc0007cf190?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:190 +0x1b4
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c970?, 0xc000f20180?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:242 +0x775
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c8d0?, 0xc0007cf4b0?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:190 +0x1b4
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c970?, 0xc000f207f8?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:242 +0x775
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c880?, 0xc00039d110?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:190 +0x1b4
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c970?, 0xc000f20858?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:242 +0x775
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c880?, 0xc00032f1f0?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:191 +0x1d3
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c948?, 0xc000599380?}, 0x1)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:237 +0x65d
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c880?, 0xc000507030?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:230 +0x5ac
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c8d0?, 0xc001214100?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:202 +0x4a7
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c970?, 0xc00117ced0?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:242 +0x775
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c8d0?, 0xc001214b60?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:190 +0x1b4
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c948?, 0xc0002992c0?}, 0x1)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:237 +0x65d
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c880?, 0xc0011d3c70?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:230 +0x5ac
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c970?, 0xc000ca3368?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:242 +0x775
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c8f8?, 0xc000fe7040?}, 0x1)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:218 +0x2e5
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000216380, {0xc0c880?, 0xc000fce0e0?}, 0x0)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:230 +0x5ac
22:55:17       honnef.co/go/tools/go/ir.(*Program).needMethodsOf(0xc000216380, {0xc0c880?, 0xc000fce0e0?})
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/methods.go:154 +0x70
22:55:17       honnef.co/go/tools/go/ir.(*Package).build(0xc00103e480)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/builder.go:2524 +0x111
22:55:17       sync.(*Once).doSlow(0xc000216380?, 0xc0007e4500?)
22:55:17         GOROOT/src/sync/once.go:74 +0xc2
22:55:17       sync.(*Once).Do(...)
22:55:17         GOROOT/src/sync/once.go:65
22:55:17       honnef.co/go/tools/go/ir.(*Package).Build(...)
22:55:17         honnef.co/go/tools/go/ir/external/co_honnef_go_tools/go/ir/builder.go:2512
22:55:17       honnef.co/go/tools/internal/passes/buildir.run(0xc00053a1a0)
22:55:17         honnef.co/go/tools/internal/passes/buildir/external/co_honnef_go_tools/internal/passes/buildir/buildir.go:86 +0x368
22:55:17       main.(*action).execOnce(0xc0001a0f30)
22:55:17         main/external/io_bazel_rules_go/go/tools/builders/nogo_main.go:402 +0x842
22:55:17       sync.(*Once).doSlow(0x0?, 0xc001b10f80?)
22:55:17         GOROOT/src/sync/once.go:74 +0xc2
22:55:17       sync.(*Once).Do(...)
22:55:17         GOROOT/src/sync/once.go:65
22:55:17       main.(*action).exec(0x0?)
22:55:17         main/external/io_bazel_rules_go/go/tools/builders/nogo_main.go:346 +0x3d
22:55:17       main.execAll.func1(0x0?)
22:55:17         main/external/io_bazel_rules_go/go/tools/builders/nogo_main.go:340 +0x54
22:55:17       created by main.execAll
22:55:17         main/external/io_bazel_rules_go/go/tools/builders/nogo_main.go:338 +0x47
```

Seems to repro in CI only in the Unused Lint, which builds all the tests with `nogo`. However, doing the same on my machine did not reproduce the issue.

Jira issue: CRDB-25156

Epic CRDB-17171

**Comment by rytaft** [comment]
This happened to me here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/10915900?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true

**Comment by rickystewart** [comment]
@rytaft I *think* this is something that will be fixed by a rebase. It looks like your branch does not contain #106256 which is the latest attempt to fix #99988.

Also, thanks for the reminder that this issue duplicates #99988. I'm going to close this issue in favor of that one (feel free to move the conversation there for any follow-ups).
